src/dune_rules/which.ml

```diff
@@ -4,19 +4,42 @@ open Memo.O
 let programs_for_which_we_prefer_opt_ext =
   [ "ocamlc"; "ocamldep"; "ocamlmklib"; "ocamlobjinfo"; "ocamlopt" ]
 
-let best_path ~dir program =
-  let exe_path program =
-    let fn = Path.relative dir (program ^ Bin.exe) in
-    let+ exists = Fs_memo.file_exists (Path.as_outside_build_dir_exn fn) in
-    if exists then Some fn else None
-  in
-  if List.mem programs_for_which_we_prefer_opt_ext program ~equal:String.equal
-  then
-    let* path = exe_path (program ^ ".opt") in
-    match path with
-    | None -> exe_path program
-    | Some _ as path -> Memo.return path
-  else exe_path program
+module Best_path = struct
+  module Make
+      (Monad : Monad.S) (Fs : sig
+        val file_exists : Path.t -> bool Monad.t
+      end) =
+  struct
+    open Monad.O
+
+    let best_path ~dir program =
+      let exe_path program =
+        let fn = Path.relative dir (program ^ Bin.exe) in
+        let+ exists = Fs.file_exists fn in
+        if exists then Some fn else None
```

Comment:
This could be `Option.some_if`, but I am on the fence myself whether it makes it better.